File: src/domain/audit/AuditLogRepository.ts
import { DataSource, EntityManager, FindOptionsWhere } from "typeorm";
import { AuditActionType } from "../../contracts/enums";
import { AuditLogEntryEntity } from "./AuditLogEntry.entity";
import { PageOptions, PageResult } from "../releaseflow/ReleaseFlowRepository";

export interface AuditLogFilter {
  releaseFlowId?: string;
  taskId?: string;
  operatorId?: string;
  actionType?: AuditActionType;
}

export class AuditLogRepository {
  private readonly repo;

  constructor(ds: DataSource) {
    this.repo = ds.getRepository(AuditLogEntryEntity);
  }

  private repoFor(em?: EntityManager) {
    return em ? em.getRepository(AuditLogEntryEntity) : this.repo;
  }

  /**
   * Append a new audit log entry.
   * This is the ONLY write operation on audit log.
   * No update or delete methods are exposed (append-only constraint).
   */
  async append(
    entry: Omit<AuditLogEntryEntity, "id" | "timestamp">,
    em?: EntityManager
  ): Promise<AuditLogEntryEntity> {
    const toSave = this.repoFor(em).create(entry as AuditLogEntryEntity);
    return this.repoFor(em).save(toSave);
  }

  async findAll(
    filter: AuditLogFilter,
    page: PageOptions
  ): Promise<PageResult<AuditLogEntryEntity>> {
    const where: FindOptionsWhere<AuditLogEntryEntity> = {};
    if (filter.releaseFlowId) where.releaseFlowId = filter.releaseFlowId;
    if (filter.taskId) where.taskId = filter.taskId;
    if (filter.operatorId) where.operatorId = filter.operatorId;
    if (filter.actionType) where.actionType = filter.actionType;

    const [data, total] = await this.repo.findAndCount({
      where,
      skip: page.page * page.size,
      take: page.size,
      order: { timestamp: "DESC" },
    });
    return { data, total, page: page.page, size: page.size };
  }
}

File: src/main/java/com/wwa/deploymentagent/domain/audit/AuditLogRepository.java
package com.wwa.deploymentagent.domain.audit;

import com.wwa.deploymentagent.contracts.enums.AuditActionType;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.stereotype.Repository;

@Repository
public interface AuditLogRepository extends JpaRepository<AuditLogEntry, String> {

    Page<AuditLogEntry> findByOperatorId(String operatorId, Pageable pageable);

    Page<AuditLogEntry> findByActionType(AuditActionType actionType, Pageable pageable);

    Page<AuditLogEntry> findByReleaseFlowId(String releaseFlowId, Pageable pageable);

    Page<AuditLogEntry> findByTaskId(String taskId, Pageable pageable);
}
